### ISSUE TITLE
routerCanReuse called twice when changing the URL manually

### ISSUE THREAD

**Issue by benneq** [angular/angular#7605]
**Note: for support questions, please use one of these channels:** https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question. This repository's issues are reserved for feature requests and bug reports.

* **I'm submitting a ... **
[ x ] bug report
[ ] feature request
[ ] support request => Please do not submit support request here, see note at the top of this template.

* **Do you want to request a *feature* or report a *bug*?**
bug

* **What is the current behavior?**
When changing the URL query param by hand (within the browser's URL bar) the `routerCanReuse` method is getting called twice.

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**
Plunkr Code: http://plnkr.co/edit/aBz13I6RMXD5PkWHPrBP

1. Open the http://run.plnkr.co/nLU7DX4Iiql4saCg/#/heroes page. (this is the preview page in external window, to see and edit the URL directly)
2. (it will automatically navigate, and add `foo=bar` parameter to the URL.
3. Open browser console.
4. Change the `foo=bar` parameter and hit the enter key.
5. You'll see that Console Output (from `app/heroes/hero-list.component.ts`) from `routerCanReuse` and `routerOnReuse` methods is getting there twice every time you change the URL param.

* **What is the expected behavior?**
I'd expect the `routerCanReuse` and `routerOnReuse` methods to only be called once, since I did change the URL only once.

* **What is the motivation / use case for changing the behavior?**
It's a bug. In my case it will fire 2 server requests every time the user edits the URL manually.

* **Please tell us about your environment:**

- Angular version: 2.0.0-beta.9
- Browser: Chrome 49 (Mac OSX 10.11)
- Language: TypeScript 1.8

* **Other information**
I'm not sure if I'm using `routerCanReuse` as intended. A current workaround for my use case would be to check for equality of `componentType` in `routerCanReuse` and for equality of `params` in `routerOnReuse`. 
If this is intended behavior, please provide some example how to use it right.

**Comment by mikesnare** [comment]
I've hit this same issue, but with a resolve guard.  I have two paths that route to the same component.  One is an id-based route for viewing and the other has path components that indicate the need to create the resource.

My goal was to have the resolve detect whether I need to create or retrieve the resource and set the resolved value accordingly.  (However, in the create case I want to redirect to the ID-based path after creation in order to prevent a refresh from re-creating the thing.

If a load a new page with the create version of the url, it works fine, and one instance is created and I redirect to the id version.  But if I change the url in that same window to either go to another page for an id OR go to the create version with new path params, the resolve guard gets called twice.

I have noticed that, for me, this only happens when my resolve returns an Observable/Promise.  It's necessary to do that in my case because either the ID or the create versions of the route need to contact the server.

The basic flow for my setup is this:

```js
return Observable.create(observer => {
	if (idMode) {
		this.http.get(`/url/${id}`)
			.map(response => response.json())
			.subscribe(
				json => {
					let thing = new Thing(jsonvalues....);
					observer.next(thing);
				},
				error => {
					// skip logging it....
					this.router.navigate(['/home']);
					observer.next(null);
					observer.complete();
				},
				() => observer.complete()
			);
	} else if (createMode) {
		this.http.post('/url',
			{
				params go here...
			})
			.map(response => response.json())
			.subscribe(
				json => {
					// creation worked, go to the id based version...
					this.router.navigate(['/path', json.id]);
					// try to prevent the original create version of the component from rendering
					observer.next(false);
				},
				error => {
					// notify user, and....
					this.router.navigate(['/home']);
					observer.next(null);
					observer.complete();
				},
				() => observer.complete()
			);
	}
});
```

This same issue happens with the back/forward browser buttons.  I.e. if I navigate manually to the create route, get forwarded to the id route, then click back, the resolve guard runs twice for the create, and then twice for the ids representing the two things that got created.

**Comment by mikesnare** [comment]
@benneq FWIW I have confirmed that your bug only happens when the HashLocationStrategy is in use.  I logged a similar bug in #11520 that calls out HashLocationStrategy specifically.

**NOTE:** It seems that the order in which you provide the HashLocationStrategy matters.  Check out `boot.ts` in my version of your plunkr at http://plnkr.co/edit/VWeXJTGOerhda1PAoCwJ.  If I provide the ROUTER_PROVIDERS before the HashLocationStrategy, the bug occurs.  If I provide the ROUTER_PROVIDERS *after* the HashLocationStrategy, the bug goes away.
